### PR TITLE
 Feat [GamePlay]: add the focer script with bullet

### PR DESCRIPTION
**Pull request by DADSILVER**
I will change the description later.

If you want to test the scene use "CantinaScriptTester".

### CODE REVIEW

File: Game/Scripts/PlayerForceUseScript.cpp [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+				rotationVerticalScript->GetField<float>("RotationSensitivity")->setter(lastVerticalSensitivity / 2.0f);
+			}
+
+			if (rotationVerticalScript)
+			{
+				lastMoveSpeed = moveScript->GetField<float>("Speed")->getter();
+				moveScript->GetField<float>("Speed")->setter(lastMoveSpeed / 2.0f);
+			}
+
+			ComponentRigidBody* rigidBody = static_cast<ComponentRigidBody*>(gameObjectAttached->GetComponent(ComponentType::RIGIDBODY));
```
Comment (markmeliaa) [inline Game/Scripts/PlayerForceUseScript.cpp:112]: Line length surpassed.

File: Game/Scripts/PlayerForceUseScript.cpp [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+			rigidBody->SetKpForce(50.0f);
+			rigidBody->SetKpTorque(50.0f);
+		}
+	}
+	else if ((input->GetKey(SDL_SCANCODE_E) == KeyState::IDLE && gameObjectAttached) || currentTimeForce < 0.0f || breakForce)
+	{
+		ComponentRigidBody* rigidBody = static_cast<ComponentRigidBody*>(gameObjectAttached->GetComponent(ComponentType::RIGIDBODY));
```
Comment (markmeliaa) [inline Game/Scripts/PlayerForceUseScript.cpp:119]: Line length surpassed.

File: Source/Physics/Physics.cpp [hunk 3]
```diff
@@ -123,6 +123,26 @@ bool Physics::Raycast(const LineSegment& ray, RaycastHit& hit, GameObject* excep
...
 
+bool Physics::RaycastToTag(const LineSegment& ray, RaycastHit& hit, GameObject* exceptionGameObject, std::string tag)
+{
+	std::map<float, const GameObject*> hitGameObjects;
+
+#ifdef ENGINE
+	AddIntersectionGameObject(hitGameObjects, ray, App->GetModule<ModuleScene>()->GetSelectedGameObject());
+#endif
+	AddIntersectionQuadtree(hitGameObjects, ray, App->GetModule<ModuleScene>()->GetLoadedScene()->GetRootQuadtree());
+	AddIntersectionDynamicObjects(hitGameObjects, ray, App->GetModule<ModuleScene>()->GetLoadedScene()->GetNonStaticObjects());
```
Comment (markmeliaa) [inline Source/Physics/Physics.cpp:134]: Line length surpassed.

**Review by markmeliaa — COMMENTED** [review]
Everything worked correctly, good job David! Just some code modifications:

File: Game/Scripts/PlayerForceUseScript.cpp [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+				rotationHorizontalScript->GetField<float>("RotationSensitivity")->setter(lastHorizontalSensitivity / 2.0f);
+			}
+
+			if (rotationVerticalScript)
+			{
+				lastVerticalSensitivity = rotationVerticalScript->GetField<float>("RotationSensitivity")->getter();
+				rotationVerticalScript->GetField<float>("RotationSensitivity")->setter(lastVerticalSensitivity / 2.0f);
+			}
+
+			if (rotationVerticalScript)
```
Comment (UriMurcia) [inline Game/Scripts/PlayerForceUseScript.cpp:106]: Is this if repeated?

File: Game/Scripts/PlayerForceUseScript.cpp [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+void PlayerForceUseScript::Update(float deltaTime)
+{
+	ModuleInput* input = App->GetModule<ModuleInput>();
+	ComponentTransform* transform = static_cast<ComponentTransform*>(owner->GetComponent(ComponentType::TRANSFORM));
+
+	if (input->GetKey(SDL_SCANCODE_E) != KeyState::IDLE && !gameObjectAttached && currentTimeForce > 14.0f)
+	{
+		RaycastHit hit;
+		Ray ray(transform->GetGlobalPosition(), transform->GetGlobalForward());
+		LineSegment line(ray, App->GetModule<ModuleScene>()->GetLoadedScene()->GetRootQuadtree()->GetBoundingBox().Size().y);
```
Comment (UriMurcia) [inline Game/Scripts/PlayerForceUseScript.cpp:76]: We could maybe set the distancePointGameObjectAttached to the ray directly? So we can limit the max distance from the beging?

**Review by UriMurcia — COMMENTED** [review]
I think it's a really good first approach for the force mechanic, really good job David! Just a few comments:

File: Game/Scripts/PlayerForceUseScript.cpp (hunk 4 above)
Comment (DADSILVER) [reply Game/Scripts/PlayerForceUseScript.cpp:106]: yeah thx, this is for the move script.

File: Game/Scripts/PlayerForceUseScript.cpp (hunk 5 above)
Comment (DADSILVER) [reply Game/Scripts/PlayerForceUseScript.cpp:76]: I tried in the LineSegment but the distance passed in the func differs. So I have to work with this approach.

**Review by diegusmejia117 — APPROVED** [review]
Teseted it on game as well as play.

Works as intended. Good Job

**Pull request merged**